src/coreclr/jit/emit.cpp

```diff
@@ -6151,9 +5946,332 @@ unsigned emitter::emitEndCodeGen(Compiler* comp,
+    args.flag         = allocMemFlag;
+
+    emitCmpHandle->allocMem(&args);
+
+    codeBlock       = (BYTE*)args.hotCodeBlock;
+    codeBlockRW     = (BYTE*)args.hotCodeBlockRW;
+    coldCodeBlock   = (BYTE*)args.coldCodeBlock;
+    coldCodeBlockRW = (BYTE*)args.coldCodeBlockRW;
+
+    consBlock   = codeBlock + emitTotalHotCodeSize + roDataAlignmentDelta;
+    consBlockRW = codeBlockRW + emitTotalHotCodeSize + roDataAlignmentDelta;
+
+#else
+
+    args.hotCodeSize  = emitTotalHotCodeSize;
+    args.coldCodeSize = emitTotalColdCodeSize;
+    args.roDataSize   = emitConsDsc.dsdOffs;
+    args.xcptnsCount  = xcptnsCount;
+    args.flag         = allocMemFlag;
+
+    emitCmpHandle->allocMem(&args);
+
+    codeBlock       = (BYTE*)args.hotCodeBlock;
+    codeBlockRW     = (BYTE*)args.hotCodeBlockRW;
+    coldCodeBlock   = (BYTE*)args.coldCodeBlock;
+    coldCodeBlockRW = (BYTE*)args.coldCodeBlockRW;
+    consBlock       = (BYTE*)args.roDataBlock;
+    consBlockRW     = (BYTE*)args.roDataBlockRW;
+
+#endif
+
+#ifdef DEBUG
+    if ((allocMemFlag & CORJIT_ALLOCMEM_FLG_32BYTE_ALIGN) != 0)
+    {
+        assert(((size_t)codeBlock & 31) == 0);
+    }
+#endif
+
+    // if (emitConsDsc.dsdOffs)
+    //     printf("Cons=%08X\n", consBlock);
+
+    /* Give the block addresses to the caller and other functions here */
+
+    *codeAddr = emitCodeBlock = codeBlock;
+    *coldCodeAddr = emitColdCodeBlock = coldCodeBlock;
+    *consAddr = emitConsBlock = consBlock;
+
+    struct Param
+    {
+        emitter*  pThis;
+        unsigned  actualCodeSize;
+        bool      contTrkPtrLcls;
+        unsigned* prologSize;
+        unsigned* epilogSize;
+        BYTE*     codeBlock;
+        BYTE*     codeBlockRW;
+        BYTE*     coldCodeBlock;
+        BYTE*     coldCodeBlockRW;
+        BYTE*     consBlock;
+        INDEBUG(unsigned* instrCount;)
+    } param;
+
+    param.pThis           = this;
+    param.contTrkPtrLcls  = contTrkPtrLcls;
+    param.prologSize      = prologSize;
+    param.epilogSize      = epilogSize;
+    param.codeBlock       = codeBlock;
+    param.codeBlockRW     = codeBlockRW;
+    param.coldCodeBlock   = coldCodeBlock;
+    param.coldCodeBlockRW = coldCodeBlockRW;
+    param.consBlock       = consBlock;
+#ifdef DEBUG
+    param.instrCount = instrCount;
+#endif
+
+    bool success = comp->eeRunWithErrorTrap<Param>(
+        [](Param* pParam) {
+            pParam->actualCodeSize =
+                pParam->pThis->writeCode(pParam->contTrkPtrLcls, pParam->prologSize, pParam->epilogSize,
+                                         pParam->codeBlock, pParam->codeBlockRW, pParam->coldCodeBlock,
+                                         pParam->coldCodeBlockRW, pParam->consBlock DEBUGARG(pParam->instrCount));
+        },
+        &param);
+
+    // Ensure that any attempt to write code after this point will fail
+    writeableOffset = 0;
+    // Notify the EE that all code writing is done. It can switch off writeable code memory.
```

Comment:
Does this meaninfully shorten the windows of where the writeable mapping is open vs. just closing the writeable mapping once we are done JITing and return to VM?

I am wondering whether it would make sense to emit the code into a scratch buffer, and then open the writeable mapping just around `memcpy` of the scratch buffer to the final location.

Reply:
I have originally had the unmapping after the method jitting was completed, but I have thought it may be better to put it right after the code is written, so I've modified it this way. 
Using a scratch buffer sounds like an interesting idea that I'd like to try. So l think I'll update this PR by removing the extraction of the writeCode method and the doneWritingCode method. It is a no-op in the current state anyways, so I can possibly put it back in a later stage.

Reply:
I did that update. @AndyAyersMS I have removed the runWithErrorTrap and doneWritingCode again based on Jan's idea.

Reply:
Fine by me, having the runtime own the cleanup makes sense.